server/service/src/pack_unit.rs

```diff
@@ -0,0 +1,176 @@
+use std::collections::HashMap;
+
+use repository::{
+    EqualFilter, PackUnitRow, PackUnitRowRepository, RepositoryError, StockLineFilter,
+    StockLineRepository, StockOnHandFilter, StockOnHandRepository, StorageConnection,
+};
+
+#[derive(Debug, Eq, PartialEq, Ord, PartialOrd)]
+pub struct PackUnit {
+    pub item_id: String,
+    pub most_used_pack_unit_id: String,
+    pub pack_units: Vec<PackUnitRow>,
+}
+
+pub fn get_pack_units(
+    connection: &StorageConnection,
+    store_id: &str,
+) -> Result<Vec<PackUnit>, RepositoryError> {
+    let stock_lines = StockLineRepository::new(connection)
+        .query_by_filter(StockLineFilter::new(), Some(store_id.to_string()))?;
+    let stock_on_hand = StockOnHandRepository::new(connection).query(Some(
+        StockOnHandFilter::new().store_id(EqualFilter::equal_to(store_id)),
+    ))?;
+    let pack_units = PackUnitRowRepository::new(connection).load_all()?;
+
+    // Calculate the most used variant for each item and pack size by total number of packs
+    // if item has stock on hand else by empty line count. HashMap keys are (item_id, pack_size)
+    // for easier management of totals.
+    let mut total_number_of_packs: HashMap<(String, i32), f64> = HashMap::new();
+    let mut total_number_of_lines: HashMap<(String, i32), f64> = HashMap::new();
+    for stock_line in stock_lines {
+        let item_id = stock_line.stock_line_row.item_id.clone();
+        let pack_size = stock_line.stock_line_row.pack_size;
+
+        if stock_on_hand
+            .iter()
+            .any(|s| s.item_id == item_id && s.available_stock_on_hand != 0)
+        {
+            total_number_of_packs
+                .entry((item_id, pack_size))
+                .and_modify(|e| *e += stock_line.stock_line_row.total_number_of_packs)
+                .or_insert(0.0);
+        } else {
+            total_number_of_lines
+                .entry((item_id, pack_size))
+                .and_modify(|e| *e += 1.0)
+                .or_insert(0.0);
+        }
+    }
+
+    // Find the most used variant id for each item based on total number of packs (if SOH) or total number of lines,
+    // and group all the pack units with their item
+    let mut pack_units_grouped: HashMap<String, (String, Vec<PackUnitRow>)> = HashMap::new();
+    for pack_unit in pack_units.clone() {
+        let item_id = pack_unit.item_id.clone();
+
+        let most_used_pack_unit_id = pack_units
+            .iter()
+            .filter(|p| p.item_id == item_id)
+            .max_by(|a, b| {
```

Comment:
Found it very hard to follow, I restructured it a bit and wonder if this is better:

```rust
let ordering_by_packs = total_number_of_packs
                    .get(&(a.item_id.clone(), a.pack_size))
                    .unwrap_or(&0.0)
                    .partial_cmp(
                        total_number_of_packs
                            .get(&(b.item_id.clone(), b.pack_size))
                            .unwrap_or(&0.0),
                    )
                    .unwrap();
                let ordering_by_lines = total_number_of_lines
                    .get(&(a.item_id.clone(), a.pack_size))
                    .unwrap_or(&0.0)
                    .partial_cmp(
                        total_number_of_lines
                            .get(&(b.item_id.clone(), b.pack_size))
                            .unwrap_or(&0.0),
                    )
                    .unwrap();

                // order by number of pack and if equal order by number of lines:
                ordering_by_packs.then(ordering_by_lines)
```

Reply:
But we want to get the max value from those HashMaps 🤔